Sources/SQLiteNIO/SQLiteDataConvertible.swift

```diff
@@ -126,7 +126,10 @@ extension Date: SQLiteDataConvertible {
         default:
             return nil
         }
-        self.init(timeIntervalSince1970: value)
+        // Round to microseconds to avoid nanosecond precision error causing Dates to fail equality
+        let valueSinceReferenceDate = value - Date.timeIntervalBetween1970AndReferenceDate
+        let microsecondsSinceReference = round(valueSinceReferenceDate * 1e6) / 1e6
+        self.init(timeIntervalSinceReferenceDate: microsecondsSinceReference)
     }
 
     public var sqliteData: SQLiteData? {
```

Comment:
Shouldn't we be rounding dates going in both directions? If we're effectively only able to load 6 digits worth of precision, ostensibly we should be unable to _store_ more than six digits either. It's true that PostgresNIO's old API (`PostgresDataConvertible`) only bothers for decoding, but @fabianfett's considerably improved `PostgresCodable` API explicitly deals in microseconds in both cases: https://github.com/vapor/postgres-nio/blob/main/Sources/PostgresNIO/New/Data/Date%2BPostgresCodable.swift

Don't forget to update the unit tests to check for correctness on both code paths 🙃 

(P.S.: Yes, Fabian, I know this is the kind of thing that makes you want to deprecate the old ones 🤣)

Reply:
I was unfortunately not able to implement this one. The problem is that the equality issue manifests because internally `Date` converts unix epoch seconds to a reference date from January 1, 2001 and visa versa. On the encoding side, `Date.sqliteData` uses `Date.timeIntervalSince1970` which adds the time between the unix epoch and the reference date which results in an extra decimal digit in these cases. While this feels disjointed, on the storing side SQLite handles this situation fine. So we are at least better off in the sense that now dates compare equal.

What do you think @gwynne? Since we can't change the to using `Date.timeIntervalSinceReferenceDate` on the encode side without breaking existing users this feels like the best we can do for now.

Reply:
Given the exact inputs you specified in your new test, I was able to more directly examine the underlying issue:
```swift
  1> import Foundation
  2> Date.timeIntervalBetween1970AndReferenceDate
$R0: Double = 978307200
  3> Date(timeIntervalSinceReferenceDate: 689658914.293192)
$R1: Foundation.Date = 2022-11-09 03:55:14 UTC
  4> Date(timeIntervalSinceReferenceDate: 689658914.293192).timeIntervalSince1970
$R2: TimeInterval = 1667966114.2931919
  5> Date(timeIntervalSinceReferenceDate: 689658914.293192).timeIntervalSinceReferenceDate
$R3: TimeInterval = 689658914.29319203
  6> Date(timeIntervalSinceReferenceDate: 689658914.2931919).timeIntervalSinceReferenceDate
$R4: TimeInterval = 689658914.29319191
```
Which ultimately comes down to meaning it's a combination of the precision of `Double` and Tanner's choice of which representation to use - and technically, fractional seconds since either reference point isn't a supported representation at all, to add insult to injury - per https://sqlite.org/datatype3.html#date_and_time_datatype, fractional values are expected to be a Julian day representation (and SQLite does its own rounding, adding even more confusion, but in practice the damage is already done before that point). As you said, however, we can't change the representation without breaking existing users, and the rounding SQLite does on incoming values means the encode is okay as it is anyhow, I think.

So, short version, you did the right thing, all good 😅.